### ISSUE TITLE
Repeated failures using hevc_nvenc

### ISSUE THREAD

**Issue by cornuwel** [Unmanic/unmanic#154]
### Description

Using the hevc_nvenc encoder with a GTX 1650, I notice a pattern: Unmanic will work for a while and then, a specific encoding will fail with that error message:

```
[hevc_nvenc @ 0x56485c7a2c80] OpenEncodeSessionEx failed: out of memory (10): (no details)
[hevc_nvenc @ 0x56485c7a2c80] No capable devices found
```

I'm not sure what's causing the error, but once it happens, all subsequent tasks will fail with the same error until I restart the container.

### Steps to Reproduce

Unclear. 
But one thing is certain, once this message is encountered, Unmanic should stop trying to encode anything since they are all guaranteed to fail...

### Versions

Version - 0.0.1+974ad93 

### Installation Method Used

Docker on Unraid

### Additional Information

Happy to provide more logs or answer any question that might help.

**Comment by Josh5** [comment]
I need more information for this. Can you please open a terminal inside the docker container and check that the GPU is still there (nvidia-smi).
Then run the same command on a terminal in the unraid host.

**Comment by cornuwel** [comment]
Doing. I started Unmanic and it is going to town on my library again.
Will post the details as soon as the error happens.

But the point I failed to convey in my original post was: As soon as we see this error message, we **know** all subsequent encodings are going to fail and it makes no sense for the worker to keep trying (and blacklisting) more files. It should stop with a warning like:
"You selected NVENC as your encoding method but we can't see your GPU. Unmanic is unable to do its job in that state so it's going to sleep until you fix it."

**Comment by cornuwel** [comment]
Damn, it finished its backlog without error, we'll have to wait a while to reproduce the problem...

I did some testing while Unmanic was running and Jellyfin was unable to transcode a video to my TV (player error). But as soon as Unmanic was done, Jellyfin started working again, without restarting anything. And it is able to do multiple transcoding at the same time for multiple clients...

I'm assuming something is going wrong with the way Unraid allocates GPUs to containers. Something around the line of "you can both have it but only one of you can use it at any given time"?
Maybe the error happens when the GPU is already working for Jellyfin and Unmanic finds a new file to transcode. This is pure conjecture at this point, but it kinda makes sense. What's weird is that Jellyfin is able to recover, whereas Unmanic is not. 
What I know for sure is that when it happens, restarting only Unmanic doesn't fix the issue. I have to stop both containers and restart Unmanic first, then Jellyfin. Something along the line of releasing the GPU completely from docker doing some sort of reset?

An ugly workaround for my use case could be to only have one "GPU container" running at any given time. I could create a cron job that stops Jellyfin at night, starts Unmanic for a few hours, and then stops it and starts Jellyfin again for the next day.

However, I still think that when Unmanic is configured to use NVENC and it doesn't find a usable GPU, it should complain and refuse to run. Whether it's caused by a configuration mistake, docker screwing up, or a hardware failure is irrelevant. The default behavior should not be to blacklist the entire library and consider the job done.

I realize I'm asking a lot of you and my small donation doesn't remotely cover the time you spend on the issues I created. I promise to make another, more substantial one, if you manage to fix that problem ;-)

**Comment by Josh5** [comment]
I'm not worried about the donation my friend :).

With regard to the sharing of the GPU, there may be an issue there.
Unmanic's container is patched to allow infinite GPU encodes. Your Jellyfin container may not be. How many workers are you running on Unmanic?

By default your 1650 is only capable of 3 encodes at a time. Any more and it will fail. Your Jellyfin docker may be enforcing that limit while the Unmanic one does not. If you are running 3 tasks on Unmanic and Jellyfin starts a task, it could be that the driver is throwing a wobbly and causing things to fail.

I can test this myself as I have this setup with a test (offical Jellyfin docker) Jellyfin library. I will take a look and see if I can confirm these suspicions.

**Comment by cornuwel** [comment]
You might not be worried about donations but I am. Your work brings a lot of value to me, and since I'm not able to contribute in code (my own projects are eating up all my time), money is the next best thing ;-)

I'm running only one worker and the container is limited to 2 cores. Because before I limited it, whatever the number of workers, all cores were pinned at 100% when Unmanic was working anyway. Not sure this is what happens (I could be wrong) as there are not that many people who have access to my Jellyfin server... Glad you are able to reproduce that setup for yourself, this will speed up the process greatly.

I'm still waiting on the next crash to be able to give you a real nvidia-smi error message...

**Comment by Trozmagon** [comment]
I'm experiencing the exact same issues when using my GTX 1650 and "Enable HW Decoding" is selected, however I don't have the same issues using my RTX 3080, I'm still doing testing on this but it's happening even if there's only 1 worker, it will hit a certain point in the job and the RAM usage will spike and run out of memory.

**Comment by Trozmagon** [comment]
Actually I spoke too soon, it just happened with "Enable HW Decoding" not selected. The one thing I can say for sure is it's not an issue on 0.0.1 tag.

**Comment by cornuwel** [comment]
I'm not sure whether you are talking about system memory or video memory. 
Either way, I have noticed an improvement since I have reduced the RAM allocation to a few VMs, leaving more system memory available for the containers : about 12GB are available with everything running...
I don't know if this coincides with your update, I'm currently running 0.0.1+63e0bca and haven't had a single crash for almost a month (last crash 26/02). I have a little over 100 successful transcodings after that...

**Comment by Trozmagon** [comment]
I'm talking about system memory, the containers RAM usage spikes and Unraid shuts it down with this error:

- server kernel: Memory cgroup out of memory: Killed process 37376 (unmanic) total-vm:16305576kB, anon-rss:3710628kB, file-rss:0kB, shmem-rss:3312kB, UID:99 pgtables:16744kB oom_score_adj:0

The issue doesn't exist on 0.0.1 like you're using, it is only happening on versions after it "0.0.2, staging, latest".
It's very hard to diagnose as there are no errors in the unmanic logs, but I can watch it when it happens.

**Comment by Trozmagon** [comment]
I'm not caching to RAM, my cache lives on an SSD and the container is limited to 8gb so it doesn't take down the server.

**Comment by Trozmagon** [comment]
I just removed the 8gb limit on the container and ran it with the latest tag, spiked up to 12gb with 1 worker.
I've added a screenshot of MediaInfo for the file but it happens on any file.

![image](https://user-images.githubusercontent.com/18479642/112495707-0b4d2b80-8d7c-11eb-83fa-7e116e739fc0.png)
[unmanic.log](https://github.com/Josh5/unmanic/files/6205860/unmanic.log)
![image](https://user-images.githubusercontent.com/18479642/112498896-dd1d1b00-8d7e-11eb-8414-5845c7ff58e1.png)

**Comment by Trozmagon** [comment]
Sorry Cornuwel, I've just re-read your issue and have noticed that they're 2 totally different issues, I got mixed up with I saw that your error message said "out of memory" and you were running a 1650.

I'll start a new issue for this as I shouldn't be hijacking yours.

**Comment by Josh5** [comment]
@Trozmagon did you create the other issue? I cannot find it...

I've found 2 issues with memory over the past month. One with the library scanner and the other with the event monitor.

Could you please test the using the 'staging' tag.

**Comment by Josh5** [comment]
Marking issue as a PICNIC and closing
